elastic/kibana
# [Fleet] Allow to restrict namespace prefix

## Summary

Resolve https://github.com/elastic/kibana/issues/182729

That PR allow to restrict authorized namespace prefix for agent policy and package policy by space, and by the same time where agent will write data. (See datastream naming scheme to see how `namespace` is used https://www.elastic.co/guide/en/fleet/current/data-streams.html#data-streams-naming-scheme)

For this I introduce a new space settings API and saved object.

That saved object is retrieved and used to validate the user configure a valid namespace each time an agent/package policy is create or updated.

### Tests

Most of the changes should be covered with automated test.

You can manually test the feature by:
* enabling the `useSpaceAwareness` feature flag
* Create a `test` space 
* Add a restriction

```
PUT kbn:/s/test/api/fleet/space_settings
{
 "allowed_namespace_prefixes": ["test"]
}
```

* Try to create an agent policy in that space with a non allowed namespace
<img width="1022" alt="Screenshot 2024-07-11 at 12 54 11 PM" src="https://github.com/elastic/kibana/assets/1336873/81a948b4-6953-406e-95ca-260dce2af8d3">





### Details

## Saved object

Introduce a new saved object `fleet_space_settings` that is space aware (type: `single`)

## API 

```
GET kbn:/s/{spaceId}/api/fleet/space_settings
```

```
PUT kbn:/s/{spaceId}/api/fleet/space_settings
{
 "allowed_namespace_prefixes": ["test1", "test2"]
}
```


## Kibana config file

The space restriction could be configured in kibana config too, for Example.

```
xpack.fleet.spaceSettings:
  - space_id: test
    allowed_namespace_prefixes: ['test']
```


## TODO

- [x] prefix validation no special character
- [x] preconfiguration
- [x] package policy


## Not done in that PR 

* Client side validation of the namespace
* UI to manage space settings
